src/platform/utilities/api/service/index.js

```diff
@@ -0,0 +1,103 @@
+const METHODS = {
+  get: 'GET',
+  post: 'POST',
+  delete: 'DELETE',
+  patch: 'PATCH',
+};
+
+const host = '/v0';
+
+export function constructUrl(method, chunks, arg = null) {
+  let url = `${host}/${chunks.join('/')}`;
+  if (typeof arg === 'number') {
+    url += `/${arg}`;
+  } else if (arg != null && typeof arg === 'object' && method === METHODS.get) {
+    // build query param string
+    const params = [];
+    Object.entries(arg).forEach(([key, value]) => {
+      if (value != null) {
+        params.push(`${key}=${value}`);
+      }
+    });
+    if (params.length) {
+      url += `?${params.join('&')}`;
+    }
+  }
+  return url;
+}
+
+export function constructOptions(method, arg = null) {
+  const options = {
+    method,
+  };
+  if ([METHODS.post, METHODS.patch].includes(method) && arg) {
+    options.body = JSON.stringify(arg);
+  }
+  return options;
+}
+
+/**
+ * Make a request to the instance url
+ * Uses default fetch options
+ * */
+export function makeRequest([url, options]) {
+  return fetch(url, {
+    ...options,
+  });
+}
+
+/**
+ * Handle the API response.
+ * Returns JSON if everything is ok, otherwise throws errors
+ * */
+export function handleResponse(response) {
+  if (response.ok) {
+    return response.json();
+  }
+
+  throw new Error(`There was an error connecting to ${response.url}`);
+}
+
+function sendFetch(arg = null) {
+  const { method, chunks } = this;
+
+  return Promise.all([
+    constructUrl(method, chunks, arg),
+    constructOptions(method, arg),
+  ])
+    .then(makeRequest)
+    .then(handleResponse.bind(this));
+}
+
+const resourceHandler = {
+  get(obj, prop) {
+    this.chunks.push(prop); // keep accumulating the path chunks
+
+    return new Proxy(sendFetch.bind({ ...this }), resourceHandler);
```

Comment:
never used a Proxy before so LMK if this correct understanding: 

```
const service = 'somePathPartString`;

api.post.beta_registrations[service]()
```

In this example, `resourceHandler.get` is called twice- once with `prop = beta_registrations` and again with `prop = service` and on each call `sendFetch` is bound to a new object that's a copy of the current `resourceHandler`. 

why is `sendFetch` bound to a copy of `resourceHandler` and not just `this`? Is it because `resourceHandler` is shared and binding it that way creates a state for a specific api request instance?

Reply:
Can we even use Proxies in IE11?

Reply:
Yeah, JS Proxies are a bit different than traditional programming so wrapping your head around them can be tricky at first.

`sendFetch` isn't bound to a copy of `resourceHandler` - it is bound to a copy of `this`.  The [proxy constructor](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy#Syntax) looks like this:

```js
var p = new Proxy(target, handler);
```

The `resourceHandler` has a `get()` function that acts as a trap for any property-accessing that happens, and that trap has this in it:

```js
return new Proxy(sendFetch.bind({ ...this }), resourceHandler);
```

Returning a Proxy with a recursive `resourceHandler` as the handler allows us to chain as many resource chunks as we want on to a request ( `api.get.whatever.you.want.forever()`).

The state for each request gets stored in `this`.  Every time `resourceHandler` has a property accessed, it appends that property to the `this.chunks` array which is then used to construct the url.

Reply:
I think erik is right- there isn't a polyfill for Proxy that works with IE 11. :(

Reply:
I think it was my first week on the job that someone broke production because we used Proxies :-)

Reply:
That's a bummer.  No IE [support](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy#Browser_compatibility) but they do work on Edge.  Do we know how much longer we'll have to support IE?

Reply:
Indefinitely, as far I know. I vaguely remember hearing once upon a time that there are a lot of VA facilities that are running IE on older machines for Veterans to use.

I think our official browser support has something to do with the popularity of the browser, but I don't recall if there's an exception for IE because of those VA facilities.